pkg/issuer/acme/dns/webhook/webhook.go

```diff
@@ -0,0 +1,131 @@
+/*
+Copyright 2018 The Jetstack cert-manager contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package webhook
+
+import (
+	"bytes"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"fmt"
+	"github.com/golang/glog"
+	"io/ioutil"
+	"net/http"
+	"time"
+
+	"github.com/jetstack/cert-manager/pkg/issuer/acme/dns/util"
+)
+
+const (
+	maxBodyLength = 5 * 1024 * 1024
+
+	presentOperation operation = "present"
+	cleanupOperation operation = "cleanup"
+)
+
+type operation string
+
+type WebhookPayload struct {
+	Operation operation `json:"operation"`
+
+	// Identifier is the record that should have a TXT record set for, e.g. _acme-challenge.example.com
+	Identifier string `json:"identifier"`
+
+	// Key is the value that the TXT record should hold for this domain
+	Key string `json:"key"`
+
+	// Metadata is arbitrary additional metadata passed to the plugin from the Issuer resource.
+	// This may contain a reference to a secret resource, containing secret data specific to this
+	// configuration of the plugin.
```

Comment:
Thanks for the comment! I think this can be re-used as a comment on our Issuer API types too 😄